src/inbox/waiting_receiver.rs

```diff
@@ -0,0 +1,169 @@
+use std::sync::{Arc, Weak};
+use std::task::{Context, Poll, Waker};
+
+use crate::envelope::MessageEnvelope;
+use crate::inbox::rx::Receiver;
+use crate::inbox::rx::RxRefCounter;
+use crate::inbox::ActorMessage;
+
+/// A [`WaitingReceiver`] is handed out by the channel any time [`Chan::try_recv`] is called on an empty mailbox.
+///
+/// [`WaitingReceiver`] implements [`Future`] which will resolve once a message lands in the mailbox.
+pub struct WaitingReceiver<A> {
+    state: Arc<spin::Mutex<WaitingState<A>>>,
```

Comment:
This is basically a oneshot, actually. I wonder whether this could lead to any greater simplification or not

Reply:
I can try replacing it with catty if you want? We already depend on that.

Reply:
However, I think it might be possible to extract our own little oneshot implementation where the sender is always a `Weak`.

Reply:
I may attempt that once I've gotten around to do the same thing for `WaitingSender`. That one is a bit more complication to change so I didn't include it in here.

Reply:
> catty doesn't use a Weak internally but two strong references so it is not quite functionally equivalent.

I think it ends up being equivalent? Catty has the same semantics where if either side drops, it disconnects.

Reply:
> > catty doesn't use a Weak internally but two strong references so it is not quite functionally equivalent.
> 
> I think it ends up being equivalent? Catty has the same semantics where if either side drops, it disconnects.

Good point. We can do that in a follow-up, right? The more important change here is to set the API and make things like `WakeReason` an implementation detail. How we realise the implementation is then free to change!

Reply:
I agree, yea